### ISSUE TITLE
Setting up a GitHub repository for your lab

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3812]
Setting up a GitHub repository for your lab<br>
<br>
http://bit.ly/2ZSZV5z<br>
<br>
About 0 Minutes<br>
<br>
<div><div>

<section>
<div>
<div>
<div>
<div>
<p>By <span>Gergana</span> on <span>May 15, 2017</span></p>
<hr>
<div>
<center>
<img alt="Img" src="https://ourcodingclub.github.io/img/tutheadergitlab.png">
</center>
</div>
<h3>Tutorial Aims:</h3>
<h4><a href="#version"> 1. Set up a lab GitHub organisational account</a></h4>
<h4><a href="#organise"> 2. Organise your lab repository</a></h4>
<h4><a href="#etiquette"> 3. Develop a coding &amp; GitHub etiquette</a></h4>
<h4><a href="#github"> 4. Learn to use RStudio and Github together</a></h4>
<h3>What is version control?</h3>
<p>Version control allows you to keep track of your work and helps you to easily explore what changes you have made, be it data, coding scripts, or manuscripts. You are probably already doing some type of version control, if you save multiple files, such as <code>Dissertation_script_25thFeb.R</code>, <code>Dissertation_script_26thFeb.R</code>, etc. This approach will leave you with tens, if not hundreds, of similar files, it makes it rather cumbersome to directly compare different versions, and is not easy to share among collaborators. What if by the time your supervisor/co-author has finished commenting on <code>Dissertation_script_26thFeb.R</code>, you are already on <code>Dissertation_script_27thFeb.R</code>? With version control software such as <a href="https://git-scm.com/" target="blank">Git</a>, version control is much smoother and easier to implement. Using an online platform like <a href="https://github.com/" target="blank">Github</a> to store your files also means that you have an online back up of your work, so you won’t need to panic when your laptop dies or your files mysteriously disappear.</p>
<p>You can use git through the command line, or through RStudio and/or GitHub desktop. You can check out <a href="https://ourcodingclub.github.io/2017/02/27/git.html" target="_blank">our Intro to Github tutorial</a> if you want to get a bit more practice with Github, but you don’t need to have completed the first tutorial to tackle this one. We encourage you to look through the <a href="#command_line"> extra resources we have added at the end of the tutorial</a> if you are keen to use the command line for your version control. But until then, here we will be syncing RStudio and Github, so you can start using version control in minutes.</p>
<h3>What are the benefits of using GitHub as a lab?</h3>
<p>Having a GitHub repo for your lab makes it easy to keep track of collaborative and personal projects - all files necessary for certain analyses can be held together and people can add in their code, graphs, etc. as the projects develop. Each file on GitHub has a history, making it easy to explore the changes that occurred to it at different time points. You can review other people’s code, add comments on certain lines or the overall document, and suggest changes. For collaborative projects, GitHub allows you to assign tasks to different users, making it clear who is responsible for which part of the analysis. You can also ask certain users to review your code. Overall, GitHub presents a great opportunity to have an online back up copy of your work, whilst also doing version control, and it can also make it easier to discuss code, as well as write and edit code collaboratively.</p>
<h3>How does GitHub work?</h3>
<h4>The members of the organisational’s account repository each have a local copy (i.e. on their computer) of all the files in the repository. The GitHub workflow can be summaried by commit-pull-push.</h4>
<h3>Commit</h3>
<p><strong>Once you’ve saved your files, you need to commit them - this means they are ready to go up on GitHub (the online copy of the repository).</strong></p>
<h3>Pull</h3>
<p><strong>Now you need to pull, i.e. make sure you are completely up to date with the online version of the files - other people could have been working on them even if you haven’t.</strong></p>
<h3>Push</h3>
<p><strong>Once you are up to date, you can push your changes - at this point in time your local copy and the online copy of the files will be the same.</strong></p>
<p><strong>Each file on GitHub has a history, so instead of having many files like <code>Dissertation_1st_May.R</code>, <code>Dissertation_2nd_May.R</code>, you can have only one and by exploring its history, you can see what it looked at different points in time.</strong></p>
<p>For example, here is the history for a script. Obviously it took me a while to calculate those model predictions!</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/filehistory.png"> </center>
<p>You can embed this workflow within <code>RStudio</code> using projects and enabling version control for them - we will be doing that shortly in the tutorial.</p>
<h4>Get lab members to register on GitHub</h4>
<p>To get started with GitHub and version control, please get each lab member to register on the <a href="https://github.com/" target="blank">Github website</a> and download and install <a href="https://git-scm.com/downloads" target="blank">Git</a> for their operating system.</p>
<h4>Register an organisational GitHub account for the lab</h4>
<p>Once you are logged in with your personal account, you can <a href="https://github.com/organizations/new">register an organisational GitHub account for the lab </a>. The files you put on GitHub will be public (i.e. everyone can see them &amp; suggest changes, but only the people with access to the account can directly edit and add/remove files). You can also have a private organisational account on GitHub, which means that only lab members can see the files. GitHub offers free private organisational accounts for educational purposes, and you can apply for one using <a href="https://education.github.com/discount_requests/new"> this link.</a></p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/register.png"> </center>
<h4>Add lab members as members of the lab’s organisational GitHub account</h4>
<p>Once all lab members are registered on GitHub and you have set up the lab’s organisational account, you can add lab members to the list of people who have access to the lab’s account. You can then decide what sort of privileges you grant users. Users can either be members, i.e. they can see all other members, can be granted access to repositories, and can also create new teams and repositories, or they can be owners with full administrative rights to the organization and have complete access to all repositories and teams. After the admin has sent out invitations to lab members, they will receive an email notification with a link to accept them. You can use the <code>Members</code> tab to see members, add new ones, and change their rights.</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/members.png"> </center>
<h3>Organise your lab repository</h3>
<h4>What is a repository?</h4>
<p>GitHub uses repositories - you can think of a repository (<em>aka</em> a repo) as a “master folder” - a repository can have folders within it, or be just separate files. In any case, a repository is what holds all of the files associated with a certain project, or in this case a certain lab’s work.</p>
<p>To make a repository, go to <code>Repositories/New repository</code> - choose a concise and informative name that has no spaces or funky characters in it. This can be your master repo that holds together past and ongoing research, data, scripts, manuscripts. Later on you might want to have more repositories - e.g. a repository associated with a particular project that you want to make public or a project where you are actively seeking feedback from a wide audience. For now, we will focus on organising and using the lab’s main repository that holds the files for all the lab’s work. <a href="https://blog.github.com/2019-01-07-new-year-new-github/" target="_blank">Github now offers free private repositories as standard with up to three collaborators</a>, so you can choose whether you want your repository to be open to public viewing or not.</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/newrepo.png"> </center>
<p>Click on <code>Initialise repo with a README.md file</code>. It’s common practice for each repository to have a <code>README.md</code> file, which contains information about the project/lab group, what is the purpose of the repository, as well as any comments on licensing and data sources. Github understands several text formats, among which <code>.txt</code> and <code>.md</code>. <code>.md</code> stands for a file written in <a href="https://en.wikipedia.org/wiki/Markdown">Markdown</a> - you might have used Markdown before from within <code>RStudio</code> to create neatly organised reports of your code and its outputs (you can also check out our <a href="https://ourcodingclub.github.io/2016/11/24/rmarkdown-1.html">Markdown tutorial</a>). You can also use Markdown to write plain text files, for example the file you are reading now was written in Markdown.</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/newrepo2.png"> </center>
<p>You can directly edit your <code>README.md</code> file on Github by clicking on the file and then selecting <code>Edit this file</code>.</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/readme.png"> </center>
<h4>Exercise 1: Write an informative README.md file</h4>
<p>You can now write the <code>README.md</code> file for your lab’s repository. To make headings and subheadings, put hashtags before a line of text - the more hashtags, the smaller the heading will appear. You can make lists using <code>-</code> and numbers <code>1, 2, 3, etc.</code>. <strong>You can discuss the information you want to include among your lab members - here are some things you might want to consider:</strong></p>
<div><div><pre><code>- Lab name, lab members, link to website &amp; social media links, contact details

- What rights do different users have - e.g. can everyone make new folders/delete files?

- Who can add new members?

- Will there be personal folders? Can other people look in them?
</code></pre></div></div>
<h4>Exercise 2: Writing a <code>.gitignore</code> file</h4>
<p>You might have noticed a file called <code>.gitignore</code> - in this file you specify which files you want Git to ignore when users make changes and add files. Examples include temporary Word, Excel and Powerpoint files, <code>.Rproj</code> files, <code>.Rhist</code> files etc. Go to <code>Create new file</code> and write a <code>.gitignore</code> file within the main repository for the lab (not any of the folders within it). You need to call the file <code>.gitignore</code> and then add the types of files that Git should ignore on separate lines. You can make this specific to your lab’s needs, but as a start, you can copy over this code:</p>
<div><div><pre><code># Prevent users to commit their own RProject
.Rproj.user
.Rproj
# Prevent users to commit their own .RData and .Rhistory in mutual area
.RData
.Rhistory
.Rapp.history
# Temporary files
*~
~$*.doc*
~$*.xls*
*.xlk
~$*.ppt*
# Prevent mac users to commit .DS_Store files
*.DS_Store
# Prevent users to commit the README files created by RStudio
*README.html
*README_cache/
#*README_files/
</code></pre></div></div>
<h4>Exercise 3: Create folders</h4>
<p>Discuss among your lab what folders your repository will contain - some examples include: manuscripts, data, figures, scripts, scripts/users/personal_folder_your_name. To make a new folder, click on <code>Create new file</code> and add in the name of your new folder, e.g. <code>manuscripts/</code> before the file name, in this case a quick <code>README.md</code> file. When creating folders within your repo through GitHub’s website, you always need to make at least one file associated with them, you can’t just create an empty folder. Add a brief explanation of what the folder is for in the <code>README.md</code> file, scroll down and click on <code>Commit new file</code>. Add a quick message where it says <code>Create README.md file</code> in light grey text - we will cover GitHub etiquette later, but for now, when creating/editing files, it’s always a good idea to change the default message to a more precise description of what was done and who did it. Stick with the default option of <code>Commit directly to master branch</code> - we will explain branches and pull requests at a later stage of the tutorial.</p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/folder.png"> </center>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/newfolder.png"> </center>
<h4>Create a <code>github-workshop</code> folder within your main lab repo - this is where we will be working for the purposes of this workshop to practice using GitHub for version control and collaborative coding.</h4>
<p>Once all lab members are synced with GitHub through <code>RStudio</code>, everyone can make folders and add files using their <code>File Explorer</code> and once they have gone through commit-pull-push, all the added folders/files will be up on GitHub, too. <strong>Note that you can’t push empty folders.</strong></p>
<h3>GitHub etiquette</h3>
<p>It’s a good idea to define some rules on how to use the lab’s repository before we start working within it - for example what GitHub and coding etiquette should lab members be following? We can make a new <code>github-etiquette.md</code> file that outlines the rules lab members agree on.</p>
<h4>Exercise 4: Write a <code>github-etiquette.md</code> file</h4>
<p>Go to your lab’s main repository, click on <code>Create new file</code> and add <code>github-etiquette.md</code> as a file name. Remember to include the file extension <code>.md</code> - otherwise GitHub won’t know what’s the file format.</p>
<p><strong>Here is a set of sample GitHub rules:</strong></p>
<div><div><pre><code>Keep your scripts in your personal scripts folder. Don't look in other people's folders.
When working on group projects, move the script out of your personal folder to the relevant project folder, so that everyone can work on it.
Keep file paths short and sensible.
Upload your data on GitHub and use relative file paths - if the data are on your computer, and you have e.g. `data &lt;- read.csv("C:/user/Documents/bird_abundance.csv")` in your code, only you will be able to read in that data, since it's only on your computer. But if you load the data from the lab's repo, every member of the lab can use it, e.g. `data &lt;- read.csv("data/bird_abundance.csv")`, where `data` is a folder within the lab's repo.
Don't use funky characters and spaces in your file names, these cause trouble because of differences in Mac/Windows systems.
Always pull before you push in case someone has done any work since the last time you pulled - you wouldn't want anyone's work to get lost or to have to resolve many coding conflicts.
</code></pre></div></div>
<h3>Coding etiquette</h3>
<p>Similarly to a GitHub etiquette, it’s useful to agree upon a set of coding etiquette rules that lab members strive to follow, so that the code within the repo is organised and easy to use.</p>
<h4>Exercise 5: Write a <code>coding-etiquette.md</code> file</h4>
<p>Discuss what to include in your coding etiquette file among lab members and write the file - here are some ideas:</p>
<p><strong>Is there a certain script format you all could use - e.g. what are the sections each script should include (e.g. libraries, load data, functions, figures) and in what order should they appear?</strong></p>
<p><strong>What syntax etiquette will lab members aim to follow?</strong></p>
<h4>“There are only two hard things in Computer Science: cache invalidation and naming things.” - Phil Karlton</h4>
<p><strong>Script file names should be meaningful and end in <code>.R</code>, e.g.:</strong></p>
<div><div><pre><code><span>farmland_figures_Feb2016.R</span><span> </span><span># Alright.</span><span>
</span><span>farmland_figures_2017</span><span>-02-25</span><span>.R</span><span> </span><span># Some people prefer this date format.</span><span>
</span><span>yet_another_script.R</span><span> </span><span># Bad. Took me hours to find the file when I needed it one year later.</span><span>
</span></code></pre></div></div>
<p><strong>Object names should be concise and meaningful.</strong></p>
<ul>
<li>
<p>Calling your data <code>data</code> might cause problems if you are doing multiple analyses at once / don’t clean your environment, and you keep using the same object name. But if you need an overwrittable universal object and you don’t need to keep lots of objects from each step of your analysis, sticking with the same object name might be useful.</p>
</li>
<li>
<p>Long object names are annoying to type - more letters, higher chance you’ll make a typo.</p>
</li>
<li>
<p>Variable and function names should be lowercase.</p>
</li>
<li>
<p>Variable names should be nouns and function names should be verbs.</p>
</li>
</ul>
<h4>- <strong>Use an underscore to separate words within a file.</strong>
</h4>
<h4>- <strong>Use a dot to separate words within objects and functions.</strong>
</h4>
<p><strong>This way it’s clear what’s an object and what’s an external file.</strong></p>
<p>The preferred form for variable names is all lower case letters and words separated with dots (<code>variable.name</code>). Function names have lowercase letters and words are separated with dots (<code>function.name</code>).</p>
<p><strong>Spacing around infix operators (<code>=, +, -, &lt;-</code>) - should there be spaces?</strong></p>
<p><strong>Line length - how long should a line of code be?</strong></p>
<p><strong>The official convention is to limit your code to 80 characters per line.</strong> Having to continuously scroll left and write can be annoying and confusing. Also, when you publish your code on Github, the scrolly bar is all the way down, so to scroll right, you first need to scroll all the way down, scroll right, then scroll all the way up to wherever you want to be - unnecessary.</p>
<p><strong>How do you know what’s 80 characters though? RStudio can place a handy line in your editor as a reminder! Go to <code>Tools/Global Options/Code/Display/Show Margin/80 characters</code>.</strong> Sometimes it might make more sense for your code to be a bit longer than 80 characters, but in general code is easier to read if there is no need for continuous scrolling left and right.</p>
<h4>When using pipes, keep the piping operator <code>%&gt;%</code> at the end of the line and continue your pipe on a new line.</h4>
<h4>When using <code>ggplot2</code>, keep the <code>+</code> at the end of the line and continue adding on layers on a new line.</h4>
<p>Here is a before and after of a <code>ggplot2</code> figure code:</p>
<div><div><pre><code><span>plot</span><span> </span><span>&lt;-</span><span> </span><span>ggplot</span><span>()</span><span>+</span><span>geom_hline</span><span>(</span><span>yintercept</span><span>=</span><span>0</span><span>,</span><span>linetype</span><span>=</span><span>"dotted"</span><span>,</span><span>colour</span><span>=</span><span>"darkgrey"</span><span>)</span><span>+</span><span>
</span><span>geom_line</span><span>(</span><span>data</span><span>=</span><span>cwa.sub</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span>=</span><span>Season</span><span>,</span><span>y</span><span>=</span><span>Total.Concentration</span><span>),</span><span>size</span><span>=</span><span>2</span><span>,</span><span>alpha</span><span>=</span><span>0.2</span><span>)</span><span>+</span><span>
</span><span>geom_ribbon</span><span>(</span><span>data</span><span>=</span><span>preds2</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span>=</span><span>Season</span><span>,</span><span> </span><span>ymin</span><span>=</span><span>ploBT</span><span>,</span><span> </span><span>ymax</span><span>=</span><span>phiBT</span><span>),</span><span> </span><span>fill</span><span>=</span><span>"#3cd0ea"</span><span>,</span><span> </span><span>alpha</span><span>=</span><span>0.3</span><span>)</span><span>+</span><span>
</span><span>geom_line</span><span>(</span><span>data</span><span>=</span><span>preds2</span><span>,</span><span>aes</span><span>(</span><span>x</span><span>=</span><span>Season</span><span>,</span><span>y</span><span>=</span><span>Total.ConcentrationBT</span><span>),</span><span>colour</span><span>=</span><span>"#3cd0ea"</span><span>,</span><span>size</span><span>=</span><span>3</span><span>)</span><span>+</span><span>theme_bw</span><span>()</span><span>+</span><span>ylab</span><span>(</span><span>"Minimum Sea Ice Concentration"</span><span>)</span><span>+</span><span>xlab</span><span>(</span><span>"Season"</span><span>)</span><span>+</span><span>annotate</span><span>(</span><span>"text"</span><span>,</span><span>x</span><span>=</span><span>2012</span><span>,</span><span>y</span><span>=</span><span>0.4</span><span>,</span><span>label</span><span>=</span><span>paste0</span><span>(</span><span>"p = "</span><span>,</span><span>round</span><span>(</span><span>pval.cwa.sub</span><span>,</span><span>4</span><span>)),</span><span>size</span><span>=</span><span>6</span><span>)</span><span>+</span><span>theme</span><span>(</span><span>legend.title</span><span>=</span><span>element_text</span><span>(</span><span>size</span><span>=</span><span>20</span><span>,</span><span>face</span><span>=</span><span>"plain"</span><span>,</span><span>hjust</span><span>=</span><span>1</span><span>),</span><span>legend.text</span><span>=</span><span>element_text</span><span>(</span><span>size</span><span>=</span><span>18</span><span>,</span><span>angle</span><span>=</span><span>45</span><span>),</span><span>legend.position</span><span>=</span><span>"bottom"</span><span>,</span><span>legend.key</span><span> </span><span>=</span><span>element_blank</span><span>(),</span><span>axis.title.x</span><span>=</span><span>element_text</span><span>(</span><span>size</span><span>=</span><span>20</span><span>,</span><span>margin</span><span>=</span><span>margin</span><span>(</span><span>20</span><span>,</span><span>0</span><span>,</span><span>0</span><span>,</span><span>0</span><span>)),</span><span>axis.title.y</span><span>=</span><span>element_text</span><span>(</span><span>size</span><span>=</span><span>20</span><span>,</span><span>margin</span><span>=</span><span>margin</span><span>(</span><span>0</span><span>,</span><span>20</span><span>,</span><span>0</span><span>,</span><span>0</span><span>)),</span><span>axis.text</span><span>=</span><span>element_text</span><span>(</span><span>size</span><span>=</span><span>16</span><span>),</span><span>panel.grid.minor</span><span>=</span><span>element_blank</span><span>(),</span><span>panel.grid.major</span><span>=</span><span>element_blank</span><span>())</span><span>
</span><span>(</span><span>plot</span><span> </span><span>&lt;-</span><span> </span><span>ggplot</span><span>()</span><span> </span><span>+</span><span>
</span><span>geom_hline</span><span>(</span><span>yintercept</span><span> </span><span>=</span><span> </span><span>0</span><span>,</span><span> </span><span>linetype</span><span> </span><span>=</span><span> </span><span>"dotted"</span><span>,</span><span> </span><span>colour</span><span> </span><span>=</span><span> </span><span>"darkgrey"</span><span>)</span><span> </span><span>+</span><span>
</span><span>geom_line</span><span>(</span><span>data</span><span> </span><span>=</span><span> </span><span>cwa.sub</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span> </span><span>=</span><span> </span><span>Season</span><span>,</span><span> </span><span>y</span><span> </span><span>=</span><span> </span><span>Total.Concentration</span><span>),</span><span> </span><span>size</span><span> </span><span>=</span><span> </span><span>2</span><span>,</span><span> </span><span>alpha</span><span> </span><span>=</span><span> </span><span>0.2</span><span>)</span><span> </span><span>+</span><span>
</span><span>geom_ribbon</span><span>(</span><span>data</span><span> </span><span>=</span><span> </span><span>preds2</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span> </span><span>=</span><span> </span><span>Season</span><span>,</span><span> </span><span>ymin</span><span> </span><span>=</span><span> </span><span>ploBT</span><span>,</span><span> </span><span>ymax</span><span> </span><span>=</span><span> </span><span>phiBT</span><span>),</span><span> </span><span>fill</span><span> </span><span>=</span><span> </span><span>"#3cd0ea"</span><span>,</span><span> </span><span>alpha</span><span> </span><span>=</span><span> </span><span>0.3</span><span>)</span><span> </span><span>+</span><span>
</span><span>geom_line</span><span>(</span><span>data</span><span> </span><span>=</span><span> </span><span>preds2</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span> </span><span>=</span><span> </span><span>Season</span><span>,</span><span> </span><span>y</span><span> </span><span>=</span><span> </span><span>Total.ConcentrationBT</span><span>),</span><span> </span><span>colour</span><span> </span><span>=</span><span> </span><span>"#3cd0ea"</span><span>,</span><span> </span><span>size</span><span> </span><span>=</span><span> </span><span>3</span><span>)</span><span> </span><span>+</span><span>
</span><span>theme_bw</span><span>()</span><span> </span><span>+</span><span>
</span><span>labs</span><span>(</span><span>y</span><span> </span><span>=</span><span> </span><span>"Minimum Sea Ice Concentration"</span><span>,</span><span> </span><span>x</span><span> </span><span>=</span><span> </span><span>"Season"</span><span>)</span><span> </span><span>+</span><span>
</span><span>annotate</span><span>(</span><span>"text"</span><span>,</span><span> </span><span>x</span><span> </span><span>=</span><span> </span><span>2012</span><span>,</span><span> </span><span>y</span><span> </span><span>=</span><span> </span><span>0.4</span><span>,</span><span> </span><span>label</span><span> </span><span>=</span><span> </span><span>paste</span><span>(</span><span>"p = "</span><span>,</span><span> </span><span>round</span><span>(</span><span>pval.cwa.sub</span><span>,</span><span>4</span><span>)),</span><span> </span><span>size</span><span> </span><span>=</span><span> </span><span>6</span><span>)</span><span> </span><span>+</span><span>
</span><span>theme</span><span>(</span><span>legend.title</span><span> </span><span>=</span><span> </span><span>element_text</span><span>(</span><span>size</span><span> </span><span>=</span><span> </span><span>20</span><span>,</span><span> </span><span>face</span><span> </span><span>=</span><span> </span><span>"plain"</span><span>,</span><span> </span><span>hjust</span><span> </span><span>=</span><span> </span><span>1</span><span>),</span><span>
</span><span>legend.text</span><span> </span><span>=</span><span> </span><span>element_text</span><span>(</span><span>size</span><span> </span><span>=</span><span> </span><span>18</span><span>,</span><span> </span><span>angle</span><span> </span><span>=</span><span> </span><span>45</span><span>),</span><span>
</span><span>legend.position</span><span> </span><span>=</span><span> </span><span>"bottom"</span><span>,</span><span>
</span><span>legend.key</span><span> </span><span>=</span><span> </span><span>element_blank</span><span>(),</span><span>
</span><span>axis.title.x</span><span> </span><span>=</span><span> </span><span>element_text</span><span>(</span><span>size</span><span> </span><span>=</span><span> </span><span>20</span><span>,</span><span> </span><span>margin</span><span> </span><span>=</span><span> </span><span>margin</span><span>(</span><span>20</span><span>,</span><span>0</span><span>,</span><span>0</span><span>,</span><span>0</span><span>)),</span><span>
</span><span>axis.title.y</span><span> </span><span>=</span><span> </span><span>element_text</span><span>(</span><span>size</span><span> </span><span>=</span><span> </span><span>20</span><span>,</span><span> </span><span>margin</span><span> </span><span>=</span><span> </span><span>margin</span><span>(</span><span>0</span><span>,</span><span>20</span><span>,</span><span>0</span><span>,</span><span>0</span><span>)),</span><span>
</span><span>axis.text</span><span> </span><span>=</span><span> </span><span>element_text</span><span>(</span><span>size</span><span>=</span><span>16</span><span>),</span><span>
</span><span>panel.grid.minor</span><span> </span><span>=</span><span> </span><span>element_blank</span><span>(),</span><span>
</span><span>panel.grid.major</span><span> </span><span>=</span><span> </span><span>element_blank</span><span>()))</span><span>
</span></code></pre></div></div>
<p>Also, remember that putting your entire ggplot code in brackets () creates the graph and then shows it in the plot viewer. If you don’t have the brackets, you’ve only created the object, but haven’t visualized it. You would then have to call the object such that it will be displayed by just typing <code>plot</code> after you’ve created the “plot” object.</p>
<h4><strong>Code chunks</strong></h4>
<p>If your code is many, many lines (as it usually is!), it would be easier for you and everyone who might need to read and use it, to split it into different sections. To do that, add in four or more dashes after your comments - you will see a little arrow appear next to that line of code and you will be able to collapse and expand the code chunk based on your needs.</p>
<p><strong>Organise your chunks in a logical order, so that your code flows nicely.</strong></p>
<p><strong>As an example of a lab’s coding etiquette, you can check out <a href="https://teamshrub.wordpress.com/">Team Shrub</a>’s <a href="https://github.com/ourcodingclub/TeamShrub-practice/blob/master/TeamShrub_Coding_Etiquette.Rmd">coding etiquette here.</a></strong></p>
<h5>For more details you can check out our <a href="https://ourcodingclub.github.io/2017/04/25/etiquette.html">Coding Etiquette tutorial.</a>
</h5>
<h3>Learn to use RStudio and Github together</h3>
<p>We are now ready to start using our repository - first all lab members need to create a local (i.e. on their computers) copy of the repository and create an RStudio project for the lab’s repo.</p>
<p><strong>Click <code>Clone or download</code> and if you are on a Windows computer, copy the HTTPS link (that’s the one that automatically appears in the box). If you have a Mac, click <code>Use SSH</code> and copy that link.</strong></p>
<center> <img alt="Img" src="https://ourcodingclub.github.io/img/clone.png"> </center>
<p>Now open <code>RStudio</code>, click <code>File/ New Project/ Version control/ Git</code> and paste the link you copied from Github. Select a directory on your computer - that is where the “local” copy of your repository will be (the online one being on Github).</p>
<p>On some Macs, <code>RStudio</code> will fail to find Git. To fix this open the terminal and type <code>sudo mv /usr/bin/git /usr/bin/git-system</code>. Then open <code>RStudio/ Preferences/ Git/SVN</code> and change <code>Git executable:</code> to <code>/usr/local/bin/git</code>. Then restart RStudio and try again. More information can be found in the <code>README.txt</code> for the Mac git installer.</p>
<p>Once the files have finished copying across, you will notice that a few things about your <code>RStudio</code> session have changed:</p>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/project2.png"></center>
<p><strong>The working directory in the top left corner is set to your local copy of the lab’s repository.</strong> You can load in data using <code>read.csv("data/your_file.csv")</code> - this would load a <code>.csv</code> file in a folder called <code>data</code> within your lab’s repository - notice that there is no need to include the repository’s name - by setting up a RStudio project, you are already within it. Similarly, when saving files, you can specify the folder where you want them saved without the repository’s name.</p>
<p><strong>There is a <code>Git</code> tab in the top right panel of RStudio.</strong> We will be doing our version control using the options within this tab.</p>
<p><strong>All the files that were in the repository online are now on your computer as well.</strong></p>
<h3>GitHub in action</h3>
<p><strong>We will practice using GitHub as a lab by analysing data from the <a href="http://www.livingplanetindex.org/home/index">Living Planet Index.</a></strong></p>
<p><strong>The data and instructions are available from <a href="https://github.com/ourcodingclub/CC-12-git-for-labs">this repository.</a> To get us all started, it would be easier if one person from the lab downloads the data and instructions and then uploads them to our new <code>github-practice</code> folder within the lab’s GitHub repository. Once that’s done, all the other lab members can pull and then everyone will have the files.</strong></p>
<p><strong>Click on <code>Clone or download</code>, <code>Download ZIP</code> and download and unzip the files from <a href="https://github.com/ourcodingclub/CC-12-git-for-labs">here.</a> Copy the files and use your file explorer to navigate to your local copy of the lab’s repository - paste the files there. Now go back to the RStudio project you set up in RStudio earlier. You can now see the files you added under the <code>Git</code> tab - you need to commit-pull-push and then the data and instructions will be up online.</strong></p>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/upload.png"></center>
<p><strong>Tick both files - you will see an <code>A</code> appear (A is for added files, if the files were modified, it would have been M, and D is for deleted files). Now click on <code>Commit</code> and write an informative commit message - it’s useful to say who you are and what you’ve done - e.g. <code>Gergana added the LPI data</code>. Click on <code>Commit</code> - there will be a bit of loading and then you’ll get a message saying your files have been added. Click on <code>Pull</code> - someone might have been working in the repo and you want to make sure you have the most recent version before you push. Once you’ve pulled, click on <code>Push</code> - your files are now online!</strong></p>
<p><strong>Everyone else should go to the <code>RStudio</code> session where they have the lab’s GitHub repo project open. Click on <code>Pull</code> in the <code>Git</code> tab in the top right corner - you are now up to date and have the files necessary to complete the coding challenges ahead!</strong></p>
<h3>Challenges &amp; collaborative coding</h3>
<h4>Open the <code>Instructions.R</code> file and follow along - if you get stuck you can check out the <code>Answers.R</code> file. The file includes several challenges - split in small teams with each team tackling one challenge - once a team completes a challenge, they can commit, pull and push!</h4>
<p><strong>Notice that the <code>Instructions.R</code> file has different sections - you can click on <code>Edit/Folding/Collapse all</code> to see only heading titles and you can enable the outline feature as well to navigate easily to certain sections. Please run all code before the start of the challenges and then only work on your challenge, ignoring the rest!</strong></p>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/outline.png"></center>
<p>Go back to your repository on Github, where you can now see all of your files (your new plots included) online. You can click on the <code>Instructions.R</code> file which should now have been modified by several people - you can click on <code>History</code> to see what everyone has done.</p>
<p>Happy coding!</p>
<h3>Potential problems</h3>
<p>Sometimes you will see error messages as you try to commit-pull-push. Usually the error message identifies the problem and which file it’s associated with, if the message is more obscure, googling it is a good step towards solving the problem. Here are some potential problems that might arise:</p>
<h4>Code conflicts</h4>
<p>While you were working on a certain part of a script, someone else was working on it, too. When you go through commit-pull-push, GitHub will make you decide which version you want to keep. This is called a code conflict, and you can’t proceed until you’ve resolved it. You will see arrows looking like <code>&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;</code> around the two versions of the code - delete the version of the code you don’t want to keep, as well as the arrows, and your conflict should disappear.</p>
<h4>Pushing the wrong files</h4>
<p>If you accidentally push not what you intended, deleted many things (or everything!) and then pushed empty folders, you can revert your commit. You can keep reverting until you reach the point in time when everything was okay.</p>
<h3>Git in the command line</h3>
<p>Traditionally, Git uses the command line to perform actions on local Git repositories. In this tutorial we ignored the command line but it is necessary if you want more control over Git. There are several excellent introductory guides on version control using Git, e.g. <a href="http://simon-m-mudd.github.io/NMDM_book/#_version_control_with_git" target="_blank">Prof Simon Mudd’s Numeracy, Modelling and Data management guide</a>, <a href="https://swcarpentry.github.io/git-novice/" target="_blank">The Software Carpentry guide</a>, and this <a href="https://github.com/BES2016Workshop/version-control" target="_blank">guide from the British Ecological Society Version Control workshop </a>. We have also created a neat cheatsheet with some basic Git commands and how they fit into the git/github ecosystem. A couple of the commands require <a href="https://github.com/github/hub"><code>hub</code></a> a wrapper for Git that increases its functionality, but not having this won’t prevent you using the other commands:</p>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/git_cli.png"></center>
<table>
<tbody>
<tr>
<th><b>Command</b></th>
<th><b>Is <code>hub</code> required?</b></th>
<th><b>Origin</b></th>
<th><b>Destination</b></th>
<th><b>Description</b></th>
</tr>
<tr>
<td><code>git fork</code></td>
<td>Y</td>
<td>Other Github</td>
<td>Personal Github</td>
<td>Creates github repo in your personal account from a previously cloned github repo.</td>
</tr>
<tr>
<td><code>git clone git@github.com:user/repo.git</code></td>
<td>N</td>
<td>Personal Github</td>
<td>Local</td>
<td>Creates a local copy of a github repo called "repo" owned by the user "user". This can be copied from github.com.</td>
</tr>
<tr>
<td><code>git add README.md</code></td>
<td>N</td>
<td>Working Dir</td>
<td>Staging Area</td>
<td>Add "README.md" to staging area.</td>
</tr>
<tr>
<td><code>git commit -m "Message"</code></td>
<td>N</td>
<td>Staging Area</td>
<td>Local</td>
<td>Commits changes to files to the local repo with the commit message "Message".</td>
</tr>
<tr>
<td><code>git commit -a -m "Message"</code></td>
<td>N</td>
<td>Working Dir</td>
<td>Local</td>
<td>adds and commits all file changes to the local repo with the commit message "Message".</td>
</tr>
<tr>
<td><code>git pull</code></td>
<td>N</td>
<td>Personal Github</td>
<td>Local</td>
<td>Retrieve any changes from a github repo.</td>
</tr>
<tr>
<td><code>git push</code></td>
<td>N</td>
<td>Local</td>
<td>Personal Github</td>
<td>Sends commited file changes to github repo.</td>
</tr>
<tr>
<td><code>git create</code></td>
<td>Y</td>
<td>Local</td>
<td>Personal Github</td>
<td>Create a github repo with the same name as the local repo.</td>
</tr>
<tr>
<td><code>git merge</code></td>
<td>N</td>
<td>NA</td>
<td>NA</td>
<td>Merge any changes in the named branch with the current branch.</td>
</tr>
<tr>
<td><code>git checkout -b patch1</code></td>
<td>N</td>
<td>NA</td>
<td>NA</td>
<td>Create a branch called "patch1" from the current branch and switch to it.</td>
</tr>
</tbody>
</table>
<hr>
<hr>
<h3><a href="https://www.surveymonkey.co.uk/r/2P9R58J" target="_blank">  We would love to hear your feedback, please fill out our survey!</a></h3>
<h3>  You can contact us with any questions on <a href="mailto:ourcodingclub@gmail.com?Subject=Tutorial%20question" target="_top">ourcodingclub@gmail.com</a>
</h3>
<h3>  Related tutorials:</h3>
<ul>
<li>
<a href="http://ourcodingclub.github.io/2018/12/02/tidyverse.html">
  - Advanced data manipulation and visualisation
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/03/06/tidyverse.html">
  - GitHub, Tidyverse and Markdown - a coding workshop for the EVENET network
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/01/22/mcmcglmm.html">
  - Meta-analysis for biologists using MCMCglmm
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2017/11/23/tutorials.html">
  - Transferring quantitative skills among scientists
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2017/05/15/git-for-labs.html">
  - Setting up a GitHub repository for your lab
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2017/02/27/git.html">
  - Intro to Github for version control
</a>
</li>
</ul>
<h3>  Subscribe to our mailing list:</h3>
</div>
</div>
</div>
</div>
</section>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via ourcodingclub.github.io http://bit.ly/2uAMVVw<br>
<br>
May 4, 2019 at 10:33PM
